File: graphs/reducers/flatten-profiles.js
// Deterministic flattening of competitor profiles into table rows (objects -> readable cells).
// input: { items: [profile] }  profile = { name, url, pricing{...}, product{...}, news{recent_news[]}, lanes{...} }
export default async function flattenProfiles(input) {
  const items = Array.isArray(input.items) ? input.items : [];
  const rows = items.map((p) => ({
    name: p?.name ?? "?",
    url: p?.url ?? "",
    pricing_model: p?.pricing?.pricing_model ?? (p?.lanes?.pricing === "completed" ? "unknown" : `(${p?.lanes?.pricing ?? "missing"})`),
    price_points: Array.isArray(p?.pricing?.price_points) ? p.pricing.price_points.slice(0, 4).join("; ") : "",
    free_tier: p?.pricing?.free_tier ?? "unknown",
    target_segment: p?.product?.target_segment ?? (p?.lanes?.product === "completed" ? "unknown" : `(${p?.lanes?.product ?? "missing"})`),
    differentiator: p?.product?.claimed_differentiator ?? "",
    key_features: Array.isArray(p?.product?.key_features) ? p.product.key_features.slice(0, 5).join("; ") : "",
    recent_news: Array.isArray(p?.news?.recent_news) ? p.news.recent_news.map((n) => `${n.date}: ${n.headline}`).slice(0, 3).join(" · ") : "",
    lanes_ok: p?.lanes ? Object.entries(p.lanes).filter(([, v]) => v === "completed").length + "/" + Object.keys(p.lanes).length : "",
  }));
  return { items: rows, fields: ["name", "pricing_model", "free_tier", "price_points", "target_segment", "differentiator", "recent_news", "lanes_ok"], _stats: { in: items.length, out: rows.length } };
}

File: kit/graphs/reducers/flatten-profiles.py
# Deterministic flattening of competitor profiles into table rows (objects -> readable cells).
# input: { items: [profile] }  profile = { name, url, pricing{...}, product{...}, news{recent_news[]}, lanes{...} }


def _lane(p, lane, value):
    lanes = p.get("lanes") or {}
    if value is not None:
        return value
    return "unknown" if lanes.get(lane) == "completed" else f"({lanes.get(lane) or 'missing'})"


def reduce(input, args, ctx):
    items = input.get("items") if isinstance(input.get("items"), list) else []
    rows = []
    for p in items:
        p = p or {}
        pricing, product, news, lanes = p.get("pricing") or {}, p.get("product") or {}, p.get("news") or {}, p.get("lanes") or {}
        rows.append({
            "name": p.get("name") or "?", "url": p.get("url") or "",
            "pricing_model": _lane(p, "pricing", pricing.get("pricing_model")),
            "price_points": "; ".join(str(x) for x in pricing["price_points"][:4]) if isinstance(pricing.get("price_points"), list) else "",
            "free_tier": pricing.get("free_tier") if pricing.get("free_tier") is not None else "unknown",
            "target_segment": _lane(p, "product", product.get("target_segment")),
            "differentiator": product.get("claimed_differentiator") or "",
            "key_features": "; ".join(str(x) for x in product["key_features"][:5]) if isinstance(product.get("key_features"), list) else "",
            "recent_news": " · ".join(f"{n.get('date')}: {n.get('headline')}" for n in news["recent_news"][:3] if isinstance(n, dict)) if isinstance(news.get("recent_news"), list) else "",
            "lanes_ok": f"{sum(1 for v in lanes.values() if v == 'completed')}/{len(lanes)}" if lanes else "",
        })
    return {"items": rows, "fields": ["name", "pricing_model", "free_tier", "price_points", "target_segment", "differentiator", "recent_news", "lanes_ok"], "_stats": {"in": len(items), "out": len(rows)}}
